[Hang Zelin] ip

HangZelin opened the pull request:
### Duke
> *“You can do anything, but not everything.” – David Allen* [(source)](https://dansilvestre.com/productivity-quotes/)

Duke frees you from worry of taking care of everything. You can ask Duke to help you remember things. It's
* Text-based!
* Easy to understand!
* Super Fast to use! :grin:

All you need is: :wink:
   1. Click this [link](https://github.com/HangZelin/ip) to **download**!
   2. Double-click to open it.
   3. Tell Duke tasks to add.
   4. Let Duke manage your tasks for you!
  
It is **FREE** to use!!!

Features:
- [x] Managing tasks: deadline, event, todo. 
- [x] Marking finished tasks.
- [x] Deleting certain tasks.
- [x] Searching certain tasks by key word or time.  
- [ ] Recognizing time for tasks (Not all format supported)
- [ ] Reminding users when ddl for certain task is approaching. (Coming soon)

****
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
    public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

HangZelin committed:
Level-1

HangZelin committed:
Level-2

HangZelin committed:
Level-3

HangZelin committed:
Level-4

HangZelin committed:
A-TextUiTesting

HangZelin committed:
Level-5

HangZelin committed:
Level-6

HangZelin committed:
A-Enums

HangZelin committed:
A-Enums

HangZelin committed:
Level-7

HangZelin committed:
Level-8

HangZelin committed:
A-MoreOOP

HangZelin committed:
A-MoreOOP

HangZelin committed:
A-Packages

HangZelin committed:
A-JUnit

HangZelin committed:
A-JUnit

HangZelin committed:
A-Jar

HangZelin committed:
A-JavaDoc

HangZelin committed:
A-CodingStandard

HangZelin committed:
Level-9

HangZelin committed:
updates for some coding standards

HangZelin committed:
Update coding Standards

HangZelin committed:
Update coding-standard

HangZelin committed:
Merge branch 'branch-Level-7'

HangZelin committed:
Merge branch 'branch-Level-8'

HangZelin committed:
Merge branch 'branch-A-CodingStandard'

HangZelin committed:
Merge branch 'branch-Level-9'

HangZelin committed:
Revert to branch-Level-7

HangZelin committed:
rebase to Level-7

HangZelin committed:
fix some indentation issues

HangZelin committed:
tiny changes

HangZelin committed:
Merge branch 'branch-Level-8'

HangZelin committed:
Some changes in Java Doc

HangZelin committed:
Merge branch 'branch-A-JavaDoc'

HangZelin committed:
delete unused jar file

HangZelin committed:
Merge branch 'master' into branch-A-CodingStandard

HangZelin committed:
Change package name

HangZelin committed:
fix merging issue

HangZelin committed:
delete Jar file

HangZelin committed:
Merge branch 'branch-A-CodingStandard'

HangZelin committed:
Merge branch 'branch-Level-9'

HangZelin committed:
Update README.md

HangZelin committed:
Update README.md

HangZelin committed:
Update README.md

HangZelin committed:
Add A-Gradle

HangZelin committed:
Merge branch 'add-gradle-support' of https://github.com/HangZelin/ip into add-gradle-support

HangZelin committed:
Update gradle plugin

HangZelin committed:
Complete checkstyle

HangZelin committed:
Set up for JavaFX

Lemonsr commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeParseException;
+
+import duke.excpetions.DukeException;
```
Perhaps check the spelling for the exceptions package here?

Lemonsr commented on src/main/java/Duke/Task/Deadlines.java:
```diff
@@ -0,0 +1,106 @@
+
+public class Deadlines extends Task {
+
+    private boolean done;
```
Should this be named isDone?

Lemonsr commented on src/main/java/Duke/SaveLoadManager/Storage.java:
```diff
@@ -0,0 +1,119 @@
+import duke.command.Parser;
+import duke.excpetions.DukeException;
+import duke.task.Task;
+import duke.task.TaskList;
```
I like how your import statements are grouped neatly here! Good job!

Lemonsr reviewed (commented):
LGTM! Just a general comment:
- Try to structure your code in a less complicated manner. (Less nested levels of if-else statements) This is just for the code to look clearer to the reader!

Overall, well done!

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+/**
+ * @author Hang Zelin
```
Its great you added author tags (i didnt haha). Seems like it should be @@author though!
https://nus-cs2103-ay2122s1.github.io/website/admin/tools.html#tool-reposense-for-authorship-tracking

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+        this.message = message;
+    }
+
+
```
Don't mean to be picky but perhaps have a consistent 1 line spacing btw methods would be good :)

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+
+
+    private static boolean isValidDate(int day, int month, int year, int hour, int minute) {
+        if (((year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)) && month == 2) {
```
Great effort in doing your own validation checks! I just used LocalDateTime haha, anyhow might be good to break it apart for readability for others and also for yourself when you relook at the code. Additional boolean variables should go a long way :)

samyipsh replied:
Also good use of guard clauses 👍

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+
+    /**
+     * Returns a LocalDateTime type that encapsulates the year, month, day, hour, minute of a time input.
+     * The method takes in a String of time and convert into LocalDateTime type.
```
~~convert into LocalDateTime type~~
converts it into a LocalDateTime

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+            int endIndex3 = time.indexOf("/", dayInteger.toString().length() + 1);
+            month = Integer.parseInt(time.substring(endIndex1 + 1, endIndex3));
+            year = Integer.parseInt(time.substring(endIndex3 + 1, endIndex2));
+
```
gd use of space to divide logic 👍

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+            return null;
+        }
+
+        //Some Other cases;
```
Perhaps best to leave out the semicolons in case it gets confused as comment out code

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+            parsedTime = LocalDate.of(year, month, day).atTime(hour, minute);
+        }
+
+
```
yup same as above, try keep spacing of 1 line consistent. Won't talk bout it again, promise 🤐

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+
+    /**
+     * Returns a String which is a task info in a local save data.
+     * Noted: you must specify it as local data, otherwise it can go wrong.
```
~~Noted~~ Note

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+    public String getSaveTask() {
+        String task;
+        char taskType = message.charAt(0);
+        if (taskType == 'D' || taskType == 'E') {
```
perhaps conversion to a switch block would be neater and more extensible.
Also don't really get the logic in 'D', 'E' block, maybe can be more explicit and not nest logic. 
If need be can always leave comments as well :)

HangZelin replied:
Yea it is just a way to better recognize save data. I will try fix that later! Really appreciate your careful look on my code! Thanks!!!

samyipsh commented on src/main/java/Duke/Task/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+     * Enum of all types of operations that is able to execute.
+     * It also contains a method AssignTask Type to find the specific type of task to create.
+     */
+    public enum OperationType {
```
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/oop.html#enumerations
Nice try using enum, might be good to CAPS everything cos its all final values

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+                || message.startsWith("find")) {
+            //Get Task description and time if it has it.
+            if (message.contains("/")) {
+                task = message.substring(message.indexOf(" ") + 1, message.indexOf("/") - 1);
```
ok, by right i should only be commenting bout code practices but i really really think youd find using regex & placing the logic of parsing into each of the subtasks much easier.

By having so many methods reliant on so many tasks, its gonna be hard to modify it.
But if you abstract the logic and put the parseLine() static method inside each task, could always just change that when you change the syntax. just fyi :)

HangZelin replied:
Yea, that will be a main stuff I will have to fix later. Thanks so much for your comments!

samyipsh commented on src/main/java/Duke/Command/Parser.java:
```diff
@@ -0,0 +1,269 @@
+                    if (message.contains("/by")) {
+                        time = message.substring(message.indexOf("/by") + 4);
+                    } else {
+                        throw new DukeException("☹ OOPS!!! I'm sorry, but the format of deadline is wrong :-(");
```
(look at comment above)
So yea, there was a statement in the website which stated that if you use more than 3 nested stuff your prolly organising your code unoptimally. Should be resolvable by doing the above suggestions.

samyipsh commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,164 @@
+
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
consider making final?

samyipsh commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,164 @@
+     */
+    public void operationForDuke(String operationType, String task, String time, int index) {
+        switch (operationType) {
+        case "bye": {
```
perhaps gd to leave out {}

samyipsh commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,164 @@
+        case "delete": {
+            try {
+                tasks.detectIndex(index);
+                ui.delete(tasks.get(index).getTaskInfo(), tasks.size() - 1);
```
perhaps can pass the handling of logical methods like delete and add to TaskList class?
Ui class sounds like it should just be handling the methods concerning interaction with user.

samyipsh commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,164 @@
+            try {
+                messages = ui.getInputForARound();
+            } catch (DukeException e) {
+                ui.printAline();
```
can abstract this double printALine() method call inside ui to a displayMessageWithFormat() method

samyipsh commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,164 @@
+
+        while (true) {
+            try {
+                messages = ui.getInputForARound();
```
maybe a btr method name could be like "returnSplitComponent" or smth like that .-.

samyipsh commented on src/main/java/Duke/Excpetions/DukeException.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public String getMessage() {
+        return this.errorMessage;
+    }
```
consider doing super(errorMessage) in constructor

then the below functions of printing out msg can be replaced by the java provided e.getMessage() function https://www.geeksforgeeks.org/3-different-ways-print-exception-messages-java/

samyipsh commented on src/main/java/Duke/SaveLoadManager/Storage.java:
```diff
@@ -0,0 +1,119 @@
+public class Storage {
+    private String filePath;
+
+    private ArrayList<Task> list;
```
consider making list final -> LIST

samyipsh commented on src/main/java/Duke/Task/Events.java:
```diff
@@ -0,0 +1,108 @@
+
+public class Events extends Task {
+
+    private boolean done;
```
isDone

samyipsh commented on src/main/java/Duke/Task/Events.java:
```diff
@@ -0,0 +1,108 @@
+    private boolean done;
+    private String task;
+    private LocalDateTime time;
+    private String taskType = "E";
```
yup so like the above, i think alot of variables can be FINAL. Id stop mentioning it

samyipsh commented on src/main/java/Duke/Task/Events.java:
```diff
@@ -0,0 +1,108 @@
+     */
+    @Override
+    public String getTaskInfo() {
+        String doneStr = "";
```
might be good to spell out String. Anyhow since im here maybe doneStatus var name could be considered.

samyipsh commented on src/main/java/Duke/Task/Events.java:
```diff
@@ -0,0 +1,108 @@
+    @Override
+    public String getTaskInfo() {
+        String doneStr = "";
+        if (!this.done) {
```
could be abstracted out into a getDoneStatus function

samyipsh commented on src/main/java/Duke/Task/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+                    ? parser.parseTime(time).format(DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm", Locale.ENGLISH))
+                    : "Nope";
+
+            if ((unParsedInfo != null && (unParsedInfo.contains(time) || unParsedInfo.contains(timeInFormat)))
```
var naming convention: ~~unParsedInfo~~ unparsedInfo

samyipsh reviewed (commented):
Sorry if I spammed the comments >< Overall, really good effort can see you really put in alot of time and tried to organise the code with meaningful segmentations and spacing. I feel your pain cos Im rly struggling with this project as well, but I really recommend you consider changing the logic of the program, might make it easier for you to extend it :)

HangZelin committed:
Complete javaFX

HangZelin committed:
Edit fxml

HangZelin committed:
Updates some JavaDoc

HangZelin committed:
Add A-Varargs

HangZelin committed:
Duke builds in gradle: extract duke as a jar file using shadowjar in gradle project.

Extracting a jar file allows users to use duke externally in PC without the IDE tool to
open up.

Using johnrengelman's shadowjar plugin, we can easily create a jar file in gradle project. After entering
"gradlew shadowJar", a jar file can be found in /build/libs, the same name as user specifies in shadowJar
config. We do so because it is a powerful feature given in shadowJar.

refer to this link to see more feature about shadowJar.

HangZelin committed:
Add Assertion Feature: Assert messages given by user will always be parsable and Duke will always response user's request.

Assertion allows program to work safely without encountering unexpected situation done by users.

Add two assert lines, one is in Duke, to ensure that Duke will always response your request, and another is on UI, to make sure
the message input can be parsed. We do so because it can satisfy our need for safe work in Duke.

HangZelin committed:
Fix some UI issues

HangZelin committed:
Merge pull request #1 from HangZelin/branch-A-Assertions

Fix some UI issues

HangZelin committed:
Resolve Github PR Reviews

HangZelin committed:
Improve code quality by coding standards.

HangZelin committed:
Merge pull request #2 from HangZelin/branch-A-CodeQuality

Branch a code quality

HangZelin committed:
Add some lambda expressions as well as optimize some codes

HangZelin committed:
Modify printList, find,  tell method to utilize java stream.

HangZelin committed:
Merge pull request #3 from HangZelin/branch-A-Streams

Branch a streams

HangZelin committed:
Set up CI

HangZelin committed:
Merge pull request #4 from HangZelin/branch-A-CI

Set up CI

HangZelin committed:
Complete Basic Undo feature

HangZelin committed:
Merge pull request #5 from HangZelin/branch-C-Undo

Complete Basic Undo feature

HangZelin committed:
Update Event and deadline to make it more SLAP style

HangZelin committed:
Add command classes to make it more OOP style, and execution class that further divides work in Duke.
Add background image and some css styles to GUI

HangZelin committed:
Update GUI UI style, add border and beautify dialog.

HangZelin committed:
Change some CI content

HangZelin committed:
Add accessibility to gradlew

HangZelin committed:
Try to Fix gradlew premission issue

HangZelin committed:
Revert "Update GUI UI style, add border and beautify dialog."

This reverts commit 0f320c60089830df48e1f51e124e6ac5bcb1809c.

HangZelin committed:
delete IO command in CI

HangZelin committed:
Merge branch 'branch-A-BetterGui' of https://github.com/HangZelin/ip into branch-A-BetterGui

HangZelin committed:
Update GUI Ui, including space between dialogs, border and a more beautiful dialog.
Delete IO CI.

HangZelin committed:
Improve some coding standards

HangZelin committed:
Try fix gradle permission error.

HangZelin committed:
Fix CI issue by solving checkstyle errors.

HangZelin committed:
Rename command.java to Command.java

HangZelin committed:
Add Display color to exception response.

HangZelin committed:
Edit some configurations

HangZelin committed:
Add some configurations to gradle build

HangZelin committed:
Fix naming issues.

HangZelin committed:
Delete src/main/java/duke/Command directory

HangZelin committed:
Delete src/main/java/duke/SaveLoadManager directory

HangZelin committed:
Delete src/main/java/duke/Task directory

HangZelin committed:
Delete src/main/java/duke/UiManager directory

HangZelin committed:
Delete same name file

HangZelin committed:
Merge pull request #7 from HangZelin/branch-A-BetterGUI

Branch a better g UI

HangZelin committed:
Delete file

HangZelin committed:
Delete src/main/java/Duke/Command directory

HangZelin committed:
Fix naming issues

HangZelin committed:
Add reuse tag

HangZelin committed:
Fix some logic bugs

HangZelin committed:
Fix some documentation, naming covention issues.

HangZelin committed:
Merge pull request #8 from HangZelin/branch-A-BetterGUI

Branch a better g UI

HangZelin committed:
Add icon and title to Duke product

HangZelin committed:
Fix some indentation issues

HangZelin committed:
Add product screenshot Ui.png

HangZelin committed:
Merge pull request #9 from HangZelin/branch-A-ProductWebsite

Add product screenshot Ui.png

HangZelin committed:
Set theme jekyll-theme-leap-day

HangZelin committed:
Fix itny issue

HangZelin committed:
Merge branch 'master' of https://github.com/HangZelin/ip into branch-A-UserGuide

HangZelin committed:
Update README.md

HangZelin committed:
Update README.md

HangZelin committed:
Update README.md

HangZelin committed:
Update README.md

HangZelin committed:
Fix some bugs in Undo

HangZelin committed:
Merge pull request #10 from HangZelin/branch-A-UserGuide

Branch a user guide

HangZelin committed:
Delete Ui.png

HangZelin committed:
Change ui title

HangZelin committed:
Add quit application

HangZelin committed:
Merge pull request #11 from HangZelin/branch-QuitApplication

Add quit application

HangZelin committed:
Modify the address of save data

HangZelin committed:
Set theme jekyll-theme-cayman

HangZelin committed:
Set theme jekyll-theme-cayman

HangZelin committed:
Set theme jekyll-theme-cayman

HangZelin committed:
Set theme jekyll-theme-cayman

HangZelin committed:
Update README.md

HangZelin committed:
Update README.md

HangZelin committed:
Update naming and fix some tiny bugs

HangZelin committed:
Change name and Ui image

HangZelin committed:
Change MANIFEST

HangZelin committed:
Add try catch block to eliminate exception issues.

HangZelin committed:
Fix more exception cases.

HangZelin committed:
Fix a few more bugs

HangZelin committed:
Fix more bugs for invalid input.

HangZelin committed:
Update README.md
